inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/AbstractKotlinElement.kt

```diff
@@ -0,0 +1,328 @@
+        return if (declaration is KSDeclaration) {
+            declaration.modifiers.contains(Modifier.JAVA_STATIC)
+        } else {
+            false
+        }
+    }
+
+    protected fun makeCopy(): AbstractKotlinElement<T> {
+        val element: AbstractKotlinElement<T> = copyThis()
+        copyValues(element)
+        return element
+    }
+
+    /**
+     * @return copy of this element
+     */
+    protected abstract fun copyThis(): AbstractKotlinElement<T>
+
+    /**
+     * @param element the values to be copied to
+     */
+    protected open fun copyValues(element: AbstractKotlinElement<T>) {
+        element.presetAnnotationMetadata = presetAnnotationMetadata
+    }
+    override fun withAnnotationMetadata(annotationMetadata: AnnotationMetadata): Element? {
+        val kotlinElement: AbstractKotlinElement<T> = makeCopy()
+        kotlinElement.presetAnnotationMetadata = annotationMetadata
+        return kotlinElement
+    }
+
+    override fun getAnnotationMetadata(): MutableAnnotationMetadataDelegate<*> {
+        if (elementAnnotationMetadata == null) {
+
+            val factory = annotationMetadataFactory
+            if (presetAnnotationMetadata == null) {
+                elementAnnotationMetadata = factory.build(this)
+            } else {
+                elementAnnotationMetadata = factory.build(this, presetAnnotationMetadata)
+            }
+        }
+        return elementAnnotationMetadata!!
+    }
+
+    override fun isPublic(): Boolean {
+        return if (declaration is KSDeclaration) {
+            declaration.getVisibility() == Visibility.PUBLIC
+        } else {
+            false
+        }
+    }
+
+    override fun isPrivate(): Boolean {
+        return if (declaration is KSDeclaration) {
+            declaration.getVisibility() == Visibility.PRIVATE
+        } else {
+            false
+        }
+    }
+
+    override fun isFinal(): Boolean {
+        return if (declaration is KSDeclaration) {
+            !declaration.isOpen() || declaration.modifiers.contains(Modifier.FINAL)
+        } else {
+            false
+        }
+    }
+
+    override fun isAbstract(): Boolean {
+        return if (declaration is KSModifierListOwner) {
+            declaration.modifiers.contains(Modifier.ABSTRACT)
+        } else {
+            false
+        }
+    }
+
+    @OptIn(KspExperimental::class)
+    override fun getModifiers(): MutableSet<ElementModifier> {
+        val dec = declaration.unwrap()
+        if (dec is KSDeclaration) {
+            val javaModifiers = visitorContext.resolver.effectiveJavaModifiers(dec)
+            return javaModifiers.mapNotNull {
+                when (it) {
+                    Modifier.ABSTRACT -> ElementModifier.ABSTRACT
+                    Modifier.FINAL -> ElementModifier.FINAL
+                    Modifier.PRIVATE -> ElementModifier.PRIVATE
+                    Modifier.PROTECTED -> ElementModifier.PROTECTED
+                    Modifier.PUBLIC -> ElementModifier.PUBLIC
```

Comment:
internal should also be public i think. but it depends on how we use getModifiers.

Reply:
yea, it has no java equivalent, but it's public in bytecode.

Reply:
ok public it is then